man/systemd.network.xml

```diff
@@ -2525,6 +2525,18 @@ NFTSet=prefix:netdev:filter:eth_ipv4_prefix</programlisting>
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>Bootp=</varname></term>
+        <listitem>
+          <para>Takes a boolean. The DHCPv4 client can be configured to communicate with BOOP servers that
+          don't accept Option 53, DHCP Message Type. In this configuration, a BOOTP Request is sent without
+          any options by default. A BOOTP reply containing Option 54: DHCP Server Identifier, and Option 1:
+          Subnet Mask is expected.</para>
```

Comment:
@colin-foster-in-advantage could you share where the reply mentioned in https://github.com/systemd/systemd/issues/34885 came from? I'm trying to figure out why it contains the Server Identifier option and why it looks like that in general. BOOTP servers aren't expected to send that option in the first place and I'm not sure why the client implemented in this PR rejects replies without that option (in other words it doesn't work with normal BOOTP servers).

The request sent by `bootpc` mentioned in https://github.com/systemd/systemd/issues/34885 is correct in the sense that it contains the magic cookie, the end option and the remaining octets of the 'vend' field are set to zero (and it isn't what is implemented here either).

Reply:
For the record I ran `bootc` to see how it reacts to the DHCP Server Identifier option. It can't recongnize it
```sh
T054='????'
```
but it accepts the reply. It accepts replies without that option just fine as well.

It even tries to guess the subnet mask if that options isn't provided
```sh
* found valid RFC1497 cookie, parsing...
* Guessing netmask from IP address range
...
```

Reply:
@evverx I'll look more into this. I don't admittedly know where the server source came from, and I don't know that I even will have access to it.

I should be able to reproduce your `bootpc` results here, at least to see that it ignores T054. I didn't know that it was ignored. From an initial look, it seems like the `lease->server_address` doesn't actually need to exist in the BOOTP response. It only applies to more DHCP-esque steps (request, renew, decline...).

I plan to remove any requirement on Option 54 in both the documentation and the unit test - you're right to point out that it doesn't make sense. I'll still be doing my validation on hardware with the server that offers Option 54, but ignores it in the same way `bootpc` does.

Does this seem reasonable?

(Also a sincere thank you to everyone helping with the reviews!)

Reply:
`bootpd` rejects the BOOTP requests sent by networkd (in its current form) with
```sh
bootpd: note(5):   received short packet
```
because the vend field is shorter than 64 bytes.

Reply:
> Does this seem reasonable?

I have to admit I'm not sure. I'm trying to figure out how it all works. Initially I didn't think that DHCP-specific options should be accepted but looking at some BOOTP servers it seems they allow sending random options. BOOTP clients apparently can send random options as well to judge from https://github.com/systemd/systemd/issues/34918 (where the Vendor class identifier was sent). I'm not sure where that BOOTP request came either. `bootpc` doesn't seem to be configurable in that regard so it must be something else.

Reply:
> I plan to remove any requirement on Option 54 in both the documentation

I think that certainly makes sense. For the record the logic behind all this I'm looking at is at https://salsa.debian.org/debian/bootpc/-/blob/master/bootpc.c?ref_type=heads with inspirational comments like
```
  /* No SUBNET TAG in the cookie so we fake guess here, if this is wrong                                                                                      
     then fix your bootp server to tell us the answer rather than                                                                                             
     hacking this code. */        
```
:-)

Reply:
> I'm not sure where that BOOTP request came either. `bootpc` doesn't seem to be configurable in that regard so it must be something else.

This one I can answer: "https://salsa.debian.org/debian/bootpc.git" with patches to fix compilation and remove automatic `strip` calls from the Makefile. No functional patch changes. The configuration for that was in `etc/network/interfaces` and was simply

```
auto lo
iface lo inet loopback

auto eth0
iface eth0 inet bootp
```

Reply:
> "https://salsa.debian.org/debian/bootpc.git"

Currently there are two BOOTP issues. As far as I understand the BOOTP request mentioned in https://github.com/systemd/systemd/issues/34885 came from that bootpc project (and it does exactly what the RFC says with the magic cookie, the end option and the zero bytes).

https://github.com/systemd/systemd/issues/34918 (with the Vendor class identifier option) is what puzzles me. I don't think `bootpc` can do that. According to https://www.rfc-editor.org/rfc/rfc2132#section-9 that option is DHCP-specific so it shouldn't be used by BOOTP clients in theory.

Reply:
@evverx I got around this by adding the following under [DHCPV4]:
```
SendOption=254:string:"UUUUUUUUUUUUUUUUUUUUUUUUUUU"
```
254 is an unused option, so this is just padding. (I chose U b/c it's 0x55, alternating 0's & 1's - easy to spot in a packet analyzer, kinda like an ethernet preamble.)

> `bootpd` rejects the BOOTP requests sent by networkd (in its current form) with
> 
> ```shell
> bootpd: note(5):   received short packet
> ```
> 
> because the vend field is shorter than 64 bytes.

Reply:
@avramd I don't think it should be necessary to manually edit the config to make networkd compliant with the RFC and get it to work with the RFC-compliant BOOTP servers.

The vend option is padded automatically in the latest version of this PR. It adds more zeroes than necessary though and the length is 68 bytes (instead of 64)

Reply:
> @evverx I'll look more into this. I don't admittedly know where the server source came from, and I don't know that I even will have access to it.

@colin-foster-in-advantage I don't think we need the BOOTP server code here, I think we need the server's configuration. Do you have access to that?